### DISCUSSION TITLE
how do I learn more about text editors, decorations, cursors, and so on?

### DISCUSSION THREAD

**Discussion by skybrian** [microsoft/vscode-discussions#755] [Extension Development QnA] [answered]
I'm writing an extension that adds a command that simulates typing into a text document like a user would, by inserting text at the cursor and moving the cursor. (It will be using an LLM, so unlike most forms of autocomplete, it will be slow enough to watch.) This could be disconcerting, so I'd like to make it clear what's going on and stop if the user does anything, including moving the cursor themselves.

I'm running up against limits in my conceptual understanding of how text editing works in VS Code. The [api docs](https://code.visualstudio.com/api) seem rather minimal and I don't find debugging into VS Code API's useful because the code is minimized. Also, browsing the source code for VS Code myself isn't easy because I don't know where anything lives and searching code on GitHub tends to find false positives. (And that's largely true of web searches as well.)

I'm learning about how things behave through experiments, but it seems like there should be an easier way. Is there documentation I'm missing? What do you do to get more in-depth knowledge?

**Comment by worksofliam** [comment]
So usually what I do is start with what APIs you might use. You already mentioned decorations, and another might be the text edit API that is part of TextDocument (I think!)

I like to search those APIs on GitHub to see how people are using them if I need examples. That helped me learn a bunch.

**Comment by r3m0t** [comment, marked as answer]
Take a look at the Monaco playground, it's like jsfiddle for Monaco editor and there are some examples.

The API in VS Code is similar, just a bit more restricted, as the editor component in VS Code shares almost all the code with Monaco.

**Reply by skybrian** [reply]
It does look interesting, but the API's seem confusingly similar with enough differences that maybe it's only useful for getting a conceptual understanding? For example, the [Position class I seem to have access to in vscode](https://github.com/microsoft/vscode/blob/afc07aa58e5d761e26db0c765d16e666fc657ada/src/vs/workbench/api/common/extHostTypes.ts#L91) has line and character fields, while the [one I see in Monaco](https://microsoft.github.io/monaco-editor/docs.html#classes/Position.html) has lineNumber and column fields. Some of the methods are different, too.

Apparently the internal and external API's are different.
